File: day-11/main.py
import statistics
import timeit
from typing import NamedTuple


class Evaluation(NamedTuple):
    # Represents an evaluation state with:
    # - number: the current number being processed
    # - num_blink: remaining blink operations
    # - factor: the number of times this state has been encountered
    number: int
    num_blink: int
    factor: float


def evaluation_loop(evaluations: list[Evaluation], num_blink: int) -> list[Evaluation]:
    # Main loop that processes evaluations for the specified number of blinks
    # Each iteration transforms the current evaluations into their children
    for _ in range(num_blink):
        evaluations = get_children_evaluations(evaluations)
    return evaluations


def get_children_evaluations(evaluations: list[Evaluation]):
    # Generates all possible child evaluations from the current set
    # Flattens the list of children and combines duplicate states
    children = [child for eval in evaluations for child in child_evaluation(eval)]
    return combine_evaluations(children)


def combine_evaluations(evaluations: list[Evaluation]) -> list[Evaluation]:
    # Combines evaluations that have the same number and num_blink
    # by adding their factors together
    combined = {}
    for number, num_blink, factor in evaluations:
        key = (number, num_blink)
        combined[key] = combined.get(key, 0) + factor
    return [Evaluation(*key, factor) for key, factor in combined.items()]


def child_evaluation(eval: Evaluation) -> list[Evaluation]:
    # Generates the next possible states for a single evaluation
    # Rules:
    # 1. If number is 0, transforms to 1
    # 2. If number has even digits, splits into two numbers
    # 3. If number has odd digits, multiplies by 2024
    number, num_blink, factor = eval

    if number == 0:
        return [Evaluation(1, num_blink - 1, factor)]

    num_str = str(number)
    if len(num_str) % 2 == 0:
        mid = len(num_str) // 2
        return [
            Evaluation(int(num_str[:mid]), num_blink - 1, factor),
            Evaluation(int(num_str[mid:]), num_blink - 1, factor),
        ]
    return [Evaluation(number * 2024, num_blink - 1, factor)]


if __name__ == "__main__":
    num_blink = 75
    initial_numbers = [41078, 18, 7, 0, 4785508, 535256, 8154, 447]
    initial_evaluations = [
        Evaluation(number, num_blink, 1) for number in initial_numbers
    ]

    def run_evaluation():
        return sum(
            eval.factor for eval in evaluation_loop(initial_evaluations, num_blink)
        )

    # Time the execution of the run_evaluation function 10 times
    times = timeit.repeat(run_evaluation, repeat=10, number=1)

    # Calculate the median of the recorded times
    median_time = statistics.median(times)

    print(f"Median execution time over 10 runs: {median_time:.6f} seconds")

File: day-11/src/main.rs
use std::collections::HashMap;
use std::time::Instant;

#[derive(Clone)]
struct Evaluation {
    number: i64,
    num_blink: i32,
    factor: f64,
}

impl Evaluation {
    #[inline(always)]
    fn new(number: i64, num_blink: i32, factor: f64) -> Self {
        Self { number, num_blink, factor }
    }
}

#[inline(always)]
fn evaluation_loop(mut evaluations: Vec<Evaluation>, num_blink: i32) -> Vec<Evaluation> {
    evaluations.reserve(evaluations.len() * 2);
    for _ in 0..num_blink {
        evaluations = get_children_evaluations(evaluations);
    }
    evaluations
}

#[inline]
fn get_children_evaluations(evaluations: Vec<Evaluation>) -> Vec<Evaluation> {
    let estimated_size = evaluations.len() * 2;
    let mut children = Vec::with_capacity(estimated_size);
    
    for eval in evaluations {
        children.extend(child_evaluation(eval));
    }
    
    combine_evaluations(children)
}

#[inline]
fn combine_evaluations(evaluations: Vec<Evaluation>) -> Vec<Evaluation> {
    let mut combined = HashMap::with_capacity(evaluations.len());
    
    for eval in evaluations {
        let key = (eval.number, eval.num_blink);
        combined.entry(key)
            .and_modify(|e| *e += eval.factor)
            .or_insert(eval.factor);
    }

    let mut result = Vec::with_capacity(combined.len());
    for ((number, num_blink), factor) in combined {
        result.push(Evaluation::new(number, num_blink, factor));
    }
    result
}

#[inline]
fn child_evaluation(eval: Evaluation) -> Vec<Evaluation> {
    let Evaluation { number, num_blink, factor } = eval;
    
    if number == 0 {
        return vec![Evaluation::new(1, num_blink - 1, factor)];
    }

    if number < 10 {
        return vec![Evaluation::new(number * 2024, num_blink - 1, factor)];
    }

    let mut n = number;
    let mut digit_count = 0;
    while n > 0 {
        digit_count += 1;
        n /= 10;
    }

    if digit_count % 2 == 0 {
        let divisor = 10_i64.pow((digit_count / 2) as u32);
        let second_half = number % divisor;
        let first_half = number / divisor;
        
        vec![
            Evaluation::new(first_half, num_blink - 1, factor),
            Evaluation::new(second_half, num_blink - 1, factor),
        ]
    } else {
        vec![Evaluation::new(number * 2024, num_blink - 1, factor)]
    }
}

fn run_calculation() -> f64 {
    let num_blink = 75;
    let initial_numbers = vec![41078, 18, 7, 0, 4785508, 535256, 8154, 447];
    
    let mut initial_evaluations = Vec::with_capacity(initial_numbers.len());
    for &number in &initial_numbers {
        initial_evaluations.push(Evaluation::new(number, num_blink, 1.0));
    }

    evaluation_loop(initial_evaluations, num_blink)
        .into_iter()
        .fold(0.0, |acc, eval| acc + eval.factor)
}

fn main() {
    let mut times = Vec::with_capacity(10);
    
    for _ in 0..10 {
        let start = Instant::now();
        run_calculation();
        let duration = start.elapsed();
        
        times.push(duration.as_micros());
    }
    
    times.sort_unstable();
    let median = if times.len() % 2 == 0 {
        (times[times.len()/2 - 1] + times[times.len()/2]) as f64 / 2.0
    } else {
        times[times.len()/2] as f64
    };
    
    println!("\nMedian execution time: {:.2} ms", median / 1000.0);
}
